Add `EthersStateManager` (#2059)
ethereumjs/ethereumjs-monorepo: packages/statemanager/README.md

await stateManager.flush()
```

### `EthersStateManager`

First, a simple example of usage:

```typescript
import { Account, Address } from '@ethereumjs/util'
import { EthersStateManager } from '@ethereumjs/statemanager'
import { ethers } from 'ethers'

const provider = new ethers.providers.JsonRpcProvider('https://path.to.my.provider.com')
const stateManager = new EthersStateManager({ provider })
const vitalikDotEth = Address.fromString('0xd8da6bf26964af9d7eed9e03e53415d37aa96045')
const account = await stateManager.getAccount(vitalikDotEth)
console.log('Vitalik has a current ETH balance of ', account.balance)
```

The `EthersStateManager` can be be used with an `ethers` `JsonRpcProvider` or one of its subclasses. Instantiate the `VM` and pass in an `EthersStateManager` to run transactions against accounts sourced from the provider or to run blocks pulled from the provider and also can compute correct updated global state roots in many instances.

Refer to [this test script](./tests/ethersStateManager.spec.ts) for complete examples of running transactions and blocks in the `vm` with data sourced from a provider.

> **WARNING** When using the `EthersStateManager`, any time an account self destruct occurs or a storage slot is deleted while running transactions and blocks, do not be surprised to see an error due to a fundamental limitation of the `web3provider` interface whereby we are not able to retrieve all of the necessary trie nodes needed to update the parent node of a deleted leaf resulting from the storage slot deletion/account self destruct. Many blocks should execute successfully and report the correct updated state root but it is not possible to predict a head of time if any particular transaction will result in this error since all keys for values in the underlying trie are hashed.

Reviewer: ```suggestion
**WARNING** When using the `EthersStateManager`, any time an account self destruct occurs or a storage slot is deleted while running transactions and blocks, do not be surprised to see an error. This is due to a fundamental limitation of the `web3provider` interface whereby we are not able to retrieve all of the necessary trie nodes needed to update the parent node of a deleted leaf resulting from the storage slot deletion/account self destruct. Many blocks should execute successfully and report the correct updated state root. However, it is not possible to predict ahead of time if any particular transaction will result in this error since all keys for values in the underlying trie are hashed.
```
slightly more readable by breaking up the sentence imo 

Concerning this part:
```
However, it is not possible to predict ahead of time if any particular transaction will result in this error since all keys for values in the underlying trie are hashed.
```
Is it actually not possible? My understanding of this is not 100% clear, but doesn't the issue only arise when the first branch-node parent of the deleted node has 2 branches (and therefore has to be converted to a leaf node/extension node when it goes from 2 branches to 1 branch)?

Author: That's precisely the scenario but there's no way to know ahead of time until you actually run the transaction and get the proof nodes into the trie that you're going to hit this scenario and even then, there's no way to recover from it since we don't have the sibling nodes to the parent node and no way to get they keys associated with those nodes.

> **WARNING** When using the `EthersStateManager`, any time an account self destruct occurs or a storage slot is deleted while running transactions and blocks, do not be surprised to see an error. This is due to a fundamental limitation of the `web3provider` interface whereby we are not able to retrieve all of the necessary trie nodes needed to update the parent node of a deleted leaf resulting from the storage slot deletion/account self destruct. Many blocks should execute successfully and report the correct updated state root. However, it is not possible to predict ahead of time if any particular transaction will result in this error since all keys for values in the underlying trie are hashed.

Reviewer 2: ```suggestion
**WARNING** When using the `EthersStateManager`, any time an account self destruct occurs or a storage slot is deleted while running transactions and blocks, do not be surprised to see an error. This is due to a fundamental limitation of the `web3provider` interface whereby we are not able to retrieve all of the necessary trie nodes needed to update the parent node of a deleted leaf resulting from the storage slot deletion/account self destruct as the keys in the underlying trie *backing* this statemanager are hashed.
However, many blocks should execute successfully and report the correct updated state root.
```
